crates/turborepo-filewatch/src/package_changes_watcher.rs

```diff
@@ -0,0 +1,226 @@
+                                    Some(new_repo_state) => {
+                                        repo_state = new_repo_state;
+                                        change_mapper = ChangeMapper::new(
+                                            &repo_state.pkg_dep_graph,
+                                            vec![],
+                                            vec![],
+                                        );
+                                    }
+                                    None => {
+                                        break;
+                                    }
+                                }
+                            }
+                            Ok(PackageChanges::Some(changed_pkgs)) => {
+                                tracing::debug!(
+                                    "changed files: {:?} changed packages: {:?}",
+                                    changed_files,
+                                    changed_pkgs
+                                );
+                                for pkg in changed_pkgs {
+                                    let _ = self.package_change_events_tx.send(
+                                        PackageChangeEvent::Package {
+                                            name: pkg.name.clone(),
+                                        },
+                                    );
+                                }
+                            }
+                            Err(err) => {
+                                // Log the error, rediscover the packages and try again
+                                tracing::error!("error: {:?}", err);
+
+                                let _ = self
+                                    .package_change_events_tx
+                                    .send(PackageChangeEvent::Rediscover);
+                                match self.initialize_repo_state().await {
+                                    Some(new_repo_state) => {
+                                        repo_state = new_repo_state;
+                                        change_mapper = ChangeMapper::new(
+                                            &repo_state.pkg_dep_graph,
+                                            vec![],
+                                            vec![],
+                                        );
+                                    }
+                                    None => {
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    Ok(Err(err)) => {
+                        tracing::error!("file event error: {:?}", err);
+                        break;
+                    }
+                    Err(broadcast::error::RecvError::Lagged(_)) => {
+                        tracing::warn!("file event lagged");
+                        break;
+                    }
```

Comment:
For these errors, are we done or do we want to loop around and try again?

Reply:
To be honest I have no idea. I erred (haha) on the side of caution. For lagging, in theory you could send a rediscover event and then continue (since the error is effectively saying we're behind on change events). I'm not sure about file errors.